src/runtime_src/core/pcie/driver/linux/xocl/subdev/address_translation.c

```diff
@@ -0,0 +1,239 @@
+ *
+ * Copyright (C) 2020 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: Chien-Wei Lan <chienwei@xilinx.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/vmalloc.h>
+#include "../xocl_drv.h"
+#include "xocl_ioctl.h"
+
+
+struct trans_addr {
+	u32 lo;
+	u32 hi;
+};
+
+#define SLBR_VER			0x0
+#define SLBR_CAP			0x4
+#define SLBR_VALID_PAGE			0x8
+#define SLBR_ENTRY_NUM			0xC
+#define SLBR_MIN_ADDR_LO		0x10
+#define SLBR_MIN_ADDR_HI		0x14
+#define SLBR_MAX_ADDR_LO		0x18
+#define SLBR_MAX_ADDR_HI		0x1C
+#define SLBR_ERR_STA			0x20
+#define SLBR_FIRST_FAIL_ADDR_LO		0x24
+#define SLBR_FIRST_FAIL_ADDR_HI		0x28
+#define SLBR_LATEST_FAIL_ADDR_LO	0x2C
+#define SLBR_LATEST_FAIL_ADDR_HI	0x30
+
+struct trans_regs {
+	u32 ver;
+	u32 cap;
+	u32 valid_page;
+	u32 entries_num;
+	struct trans_addr min_addr;
+	struct trans_addr max_addr;
+	u32 err_status;
+	struct trans_addr first_fail_addr;
+	struct trans_addr latest_fail_addr;
+	u8 padding[204];
+	struct trans_addr page_table_phys[256];
+};
+
+#define	SLBR_DEV2XDEV(d)	xocl_get_xdev(d)
+
+struct slbr {
+	void __iomem		*base;
+	struct device		*dev;
+	struct mutex		lock;
+	bool			online;
+};
+
+static ssize_t status_show(struct device *dev, struct device_attribute *attr,
+	char *buf)
+{
+	u32 status = 1;
+
+	return sprintf(buf, "0x%x\n", status);
+}
+static DEVICE_ATTR_RO(status);
+
+static struct attribute *slbr_attributes[] = {
+	&dev_attr_status.attr,
+	NULL
+};
+
+static const struct attribute_group slbr_attrgroup = {
+	.attrs = slbr_attributes,
+};
+
+static uint32_t slbr_get_entries_num(struct platform_device *pdev)
+{
+	struct slbr *slbr = platform_get_drvdata(pdev);
+	struct trans_regs *regs = (struct trans_regs *)slbr->base;
+	xdev_handle_t xdev = SLBR_DEV2XDEV(pdev);
+	uint32_t num;
+
+	mutex_lock(&slbr->lock);
+
+	num = (xocl_dr_reg_read32(xdev, &regs->cap)>>6 & 0x1f) << 8;
+
+	num = 256;
+	mutex_unlock(&slbr->lock);
+	return num;
+}
+
+
+static int slbr_set_page_table(struct platform_device *pdev, uint64_t *phys_addrs, uint64_t base_addr, uint64_t entry_sz, uint32_t num)
+{
+	int ret = 0, i = 0;
+	uint32_t num_max;
+	struct slbr *slbr = platform_get_drvdata(pdev);
+	struct trans_regs *regs = (struct trans_regs *)slbr->base;
+	xdev_handle_t xdev = SLBR_DEV2XDEV(pdev);
+	uint64_t min_addr = base_addr, max_addr = base_addr + num * entry_sz;
+
+	mutex_lock(&slbr->lock);
+
+	num_max = (xocl_dr_reg_read32(xdev, &regs->cap)>>6 & 0x1f) << 8;
+
+	if (num > num_max) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	for ( ; i < num; ++i) {
```

Comment:
do you want to check and make sure num is a power of two?

Reply:
power of 2 check added